packages/composable-controller/src/ComposableController.ts

```diff
@@ -1,75 +1,111 @@
 
-export type ComposableControllerRestrictedMessenger =
-  RestrictedControllerMessenger<'ComposableController', never, any, never, any>;
+/**
+ * List of child controller instances
+ */
+export type ControllerList = ControllerInstance[];
 
 /**
- * Controller that can be used to compose multiple controllers together.
+ * Determines if the given controller is an instance of BaseControllerV1
+ * @param controller - Controller instance to check
+ * @returns True if the controller is an instance of BaseControllerV1
  */
-export class ComposableController extends BaseControllerV1<never, any> {
-  private readonly controllers: ControllerList = [];
+function isBaseControllerV1(
+  controller: ControllerInstance,
+): controller is BaseControllerV1<any, any> {
+  return 'subscribe' in controller && controller.subscribe !== undefined;
+}
 
-  private readonly messagingSystem?: ComposableControllerRestrictedMessenger;
+export type ComposableControllerState = {
+  [name: string]: ControllerInstance['state'];
+};
 
-  /**
-   * Name of this controller used during composition
-   */
-  override name = 'ComposableController';
+export type ComposableControllerGetStateAction = ControllerGetStateAction<
+  `${typeof controllerName}`,
+  ComposableControllerState
+>;
+
+export type ComposableControllerStateChangeEvent = ControllerStateChangeEvent<
+  `${typeof controllerName}`,
+  ComposableControllerState
+>;
+
+export type ComposableControllerActions = ComposableControllerGetStateAction;
+
+export type ComposableControllerEvents = ComposableControllerStateChangeEvent;
+
+export type ComposableControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  ControllerGetStateAction<string, Record<string, unknown>>,
+  ControllerStateChangeEvent<string, Record<string, unknown>>,
+  string,
+  string
+>;
+
+/**
+ * Controller that can be used to compose multiple controllers together.
+ */
+export class ComposableController extends BaseController<
+  typeof controllerName,
+  ComposableControllerState,
+  ComposableControllerMessenger
+> {
+  readonly #controllers: ControllerList = [];
 
   /**
    * Creates a ComposableController instance.
    *
-   * @param controllers - Map of names to controller instances.
-   * @param messenger - The controller messaging system, used for communicating with BaseController controllers.
+   * @param options - Initial options used to configure this controller
+   * @param options.controllers - List of child controller instances to compose.
+   * @param options.messenger - A restricted controller messenger.
    */
-  constructor(
-    controllers: ControllerList,
-    messenger?: ComposableControllerRestrictedMessenger,
-  ) {
-    super(
-      undefined,
-      controllers.reduce((state, controller) => {
+
+  constructor({
+    controllers,
+    messenger,
+  }: {
+    controllers: ControllerList;
+    messenger: ComposableControllerMessenger;
+  }) {
+    if (messenger === undefined) {
```

Comment:
I fixed the error message so that we're always throwing an error if messenger is undefined: ac750ec7f5a38489ddbad340d2ce0322d01fa1d3. 

This matches the current test for this branch. Also, the `super({ messenger })` call in the constructor is now required, so an error will be thrown upstream if `messenger` is not provided.